src/gausskernel/storage/access/k2/pg_session.cpp

```diff
@@ -91,4 +91,33 @@ std::shared_ptr<PgTableDesc> PgSession::LoadTable(const PgObjectId& table_object
     }
 }
 
+bool operator==(const PgForeignKeyReference& k1, const PgForeignKeyReference& k2) {
+  return k1.table_oid == k2.table_oid &&
+      k1.k2pgctid == k2.k2pgctid;
+}
+
+size_t hash_value(const PgForeignKeyReference& key) {
+  size_t hash = 0;
+  boost::hash_combine(hash, key.table_oid);
+  boost::hash_combine(hash, key.k2pgctid);
+  return hash;
+}
+
+bool PgSession::ForeignKeyReferenceExists(uint32_t table_oid, std::string&& k2pgctid) {
```

Comment:
Do you know how this foreign key cache is used? I'm wondering if we need to make this distributed in k2 to work properly with distributed sql instances.

Reply:
foreign key cache is used for a tuple insert or delete, which is checked by postgres internal logic based on foreign key constraints in table/index schema. That is why the foreign key cache is only pg_session based and there is no need to distribute it to different sql instances.